contracts/core/modules/RebalancingSetExchangeIssuanceModule.sol

```diff
@@ -134,313 +139,396 @@ contract RebalancingSetExchangeIssuanceModule is
+        uint256 _rebalancingSetQuantity,
+        address _paymentTokenAddress,
+        uint256 _paymentTokenQuantity,
+        ExchangeIssuanceLibrary.ExchangeIssuanceParams memory _exchangeIssuanceParams,
+        bytes memory _orderData,
+        bool _keepChangeInVault
     )
         private
-        view
-        returns (uint256)
     {
-        uint256 rbSetUnitShares = IRebalancingSetToken(_rebalancingSetAddress).unitShares();
-        uint256 rbSetNaturalUnit = IRebalancingSetToken(_rebalancingSetAddress).naturalUnit();
+        address baseSetAddress = _exchangeIssuanceParams.setAddress;
+        uint256 baseSetIssueQuantity = _exchangeIssuanceParams.quantity;
 
-        // Calculate the possible number of Sets issuable (may not be a multiple of natural unit)
-        uint256 possibleIssuableRBSetQuantity = _baseSetIssueQuantity.mul(rbSetNaturalUnit).div(rbSetUnitShares);
+        // Validate parameters
+        validateInputs(
+            _paymentTokenAddress,
+            _rebalancingSetAddress,
+            _rebalancingSetQuantity,
+            baseSetAddress,
+            _exchangeIssuanceParams.sendTokens
+        );
+
+        // Ensure payment token allowance to the TransferProxy
+        // Note that the paymentToken may also be used as a component to issue the Set
+        ERC20Wrapper.ensureAllowance(
+            _paymentTokenAddress,
+            address(this),
+            transferProxy,
+            _paymentTokenQuantity
+        );
 
-        // Ensure that the base Set quantity is a multiple of the rebalancing Set natural unit
-        uint256 rbSetIssueQuantity = possibleIssuableRBSetQuantity.div(rbSetNaturalUnit).mul(rbSetNaturalUnit);
+        // Atomically trade paymentToken for base SetToken components
+        // and mint the base SetToken
+        exchangeIssuanceInstance.exchangeIssue(
+            _exchangeIssuanceParams,
+            _orderData
+        );
+
+        // Approve base SetToken to transfer proxy for minting rebalancing SetToken
+        ERC20Wrapper.ensureAllowance(
+            baseSetAddress,
+            address(this),
+            transferProxy,
+            baseSetIssueQuantity
+        );
+
+        // Issue rebalancing SetToken to the caller
+        coreInstance.issueTo(
+            msg.sender,
+            _rebalancingSetAddress,
+            _rebalancingSetQuantity
+        );
 
-        return rbSetIssueQuantity;
+        // Send excess base Set held in this contract to the user
+        // If keepChangeInVault is true, the baseSetToken is held in the Vault
+        // which is a UX improvement
+        returnExcessBaseSetFromContract(
+            baseSetAddress,
+            transferProxy,
+            _keepChangeInVault
+        );
+
+        // Return any extra components acquired during exchangeIssue to the user
+        returnExcessComponentsFromVault(baseSetAddress);
+
+        // Note: paymentTokenQuantity could be spoofed
+        emit LogPayableExchangeIssue(
```

Comment:
Do we want to log the quantity if it can be spoofed?

Reply:
Maybe we log the `_rebalancingSetQuantity`? Feels more relevant and like you said can't be manipulated.